### PR TITLE
Fix Heap Buffer Overflow when sending message with Property (#1618)

### PR DESCRIPTION
**Pull request by ewertons**
# Checklist
- [x] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-c/blob/master/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.
- I submitted this PR against the correct branch: 
  - [x] This pull-request is submitted against the `master` branch. 
  - [x] I have merged the latest `master` branch prior to submission and re-merged as needed after I took any feedback.
  - [x] I have squashed my changes into one with a clear description of the change.

# Reference/Link to the issue solved with this PR (if any)
https://github.com/Azure/azure-iot-sdk-c/issues/1618

# Description of the problem

# Description of the solution

### CODE REVIEW

**Review by ericwol-msft — APPROVED** [review]

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 1]
```diff
@@ -1203,15 +1203,44 @@ static int subscribeToNotifyStateIfNeeded(PMQTTTRANSPORT_HANDLE_DATA transport_d
     return result;
 }
 
+// The tokenData variable carries a string in the format "<name>=<value".
```
Comment (momuno) [inline iothub_client/src/iothubtransport_mqtt_common.c:1206]: "\<name\>=\<value" to "\<name\>=\<value\>"

**Review by momuno — APPROVED** [review]

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 2]
```diff
@@ -1203,15 +1203,44 @@ static int subscribeToNotifyStateIfNeeded(PMQTTTRANSPORT_HANDLE_DATA transport_d
...
+// This function returns the length of the <name> (the portion of the string from position 0 to the last char before '=').
+// If the '=' character is not present (unlikely), the length returned is 0, which should be
+// properly handled by the calling function.
+static int getPropertyLength(const char* tokenData)
```
Comment (ericwol-msft) [inline iothub_client/src/iothubtransport_mqtt_common.c:1210]: {
    int result = 0;

    while (tokenData[result] == '\0')
    {
        if (tokenData[result] == '=')
        {
            return result;
        }
        result++;
    }

    return 0;
}

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 3]
```diff
@@ -1203,15 +1203,44 @@ static int subscribeToNotifyStateIfNeeded(PMQTTTRANSPORT_HANDLE_DATA transport_d
...
+}
 
 static bool isSystemProperty(const char* tokenData)
 {
     bool result = false;
     size_t propCount = sizeof(sysPropList) / sizeof(sysPropList[0]);
     size_t index = 0;
+    size_t tokenDataLength = getPropertyLength(tokenData);
+
     for (index = 0; index < propCount; index++)
```
Comment (ericwol-msft) [inline iothub_client/src/iothubtransport_mqtt_common.c:1240]: for (index = 0; tokenDataLength  > 0 && index < propCount; index++)

**Review by ericwol-msft — APPROVED** [review]

File: iothub_client/src/iothubtransport_mqtt_common.c (hunk 2 above)
Comment (momuno) [reply iothub_client/src/iothubtransport_mqtt_common.c:1210]: But what if tokenData does not have a terminating null character? Could there be a seg fault?

File: iothub_client/src/iothubtransport_mqtt_common.c (hunk 3 above)
Comment (ewertons) [reply iothub_client/src/iothubtransport_mqtt_common.c:1213]: the tokenDataLength must be compared with each property in the list we maintain in code, so this check would not address the issue.

File: iothub_client/src/iothubtransport_mqtt_common.c (hunk 2 above)
Comment (ewertons) [reply iothub_client/src/iothubtransport_mqtt_common.c]: This function has been removed. Unfortunately it didnt work for all the cases, according to how the MQTT transport is implemented. See comment on the issue discussion.

File: iothub_client/src/iothubtransport_mqtt_common.c (hunk 1 above)
Comment (ewertons) [reply iothub_client/src/iothubtransport_mqtt_common.c]: Good catch. However this function has been removed.

**Comment by ewertons** [comment]
The MQTT transport in the Azure-IoT-SDK-C code optimizes the logic to search for system properties in C2D message topic names. 

Instead of looking for "%24.cid" (which is usually the first property in the topic name), the code checks if a token starts  

with "device/" (e.g., in the token "devices/csdk_e2eDevice_sas_j_please_delete_06a6b799-7867-45d8-85ba-4716cdbf15ae/messages/devicebound/%24.cid=MessageCorrelationIdForE2E") 

assuming that such token contains "cid" as first property, extracting the property value (the portion after '=') and setting the correlation id using IoTHubMessage_SetCorrelationId. 

Given that, this fix was simplified as much as possible.

**Pull request merged**